src/shared/modules/auth/helpers.js

```diff
@@ -0,0 +1,133 @@
+import { isObject } from 'lodash'
+import { AUTH_LOGGING_PREFIX, AUTH_STORAGE_LOGS } from './constants'
+import { isAuthLoggingEnabled, isAuthDebuggingEnabled } from './settings'
+import { saveAs } from 'file-saver'
+
+export const authLog = (msg, type = 'log') => {
+  if (!isAuthLoggingEnabled) return
+  if (!['log', 'error', 'warn'].includes(type)) return
+  const log = `${AUTH_LOGGING_PREFIX} [${new Date().toISOString()}] ${msg}`
+  const logs = sessionStorage.getItem(AUTH_STORAGE_LOGS) || ''
+  const logsLines = logs.split('\n')
+
+  const truncatedOldLogs =
+    logsLines.length > 200 ? logsLines.slice(-199).join('\n') : logs
+
+  sessionStorage.setItem(AUTH_STORAGE_LOGS, `${truncatedOldLogs}${log}\n`)
+}
+
+export const downloadAuthLogs = () => {
+  const blob = new Blob([sessionStorage.getItem(AUTH_STORAGE_LOGS)], {
+    type: 'text/plain;charset=utf-8'
+  })
+  saveAs(blob, 'sso.log')
+}
+
+export const authDebug = (msg, content) => {
+  if (!isAuthDebuggingEnabled) return
+  console.log(`${AUTH_LOGGING_PREFIX} - DEBUG - ${msg}`)
+  console.dir(content)
+}
+
+export const createNonce = () =>
+  Array.from(window.crypto.getRandomValues(new Uint32Array(4)), t =>
+    t.toString(36)
+  ).join('-')
+
+export const createStateForRequest = () => {
+  const base = Array.from(
+    window.crypto.getRandomValues(new Uint32Array(4)),
+    t => t.toString(16)
+  ).join('-')
+  return `state-${base}`
+}
+
+export const createCodeVerifier = method => {
+  const errorMessage = `${AUTH_LOGGING_PREFIX} code verifier, Invalid argument`
+
+  switch (method) {
+    case 'plain':
+    case 'S256':
+      const randomString = Array.from(
+        window.crypto.getRandomValues(new Uint8Array(32)),
+        t => String.fromCharCode(t)
+      ).join('')
+      return _btoaUrlSafe(randomString)
+    case '':
```

Comment:
weird, maybe there should be an early return (single line of course!) that just checks for all falsy values and returns null...

Also weird that it throws a string for the default switch case...